engine/sources/local/localfs.go

```diff
@@ -0,0 +1,808 @@
+				}
+				cachedResultsMu.Lock()
+				cachedResults = append(cachedResults, appliedChange)
+				cachedResultsMu.Unlock()
+				if cacheCtx != nil {
+					if err := cacheCtx.HandleChange(appliedChange.Result().kind, path, appliedChange.Result().stat, nil); err != nil {
+						return fmt.Errorf("failed to handle change in content hasher: %w", err)
+					}
+				}
+
+				return nil
+
+			case upperStat.Linkname != "":
+				// delay hardlinks until after everything else so we know the source of the link exists
+				hardlinkMu.Lock()
+				hardlinks = append(hardlinks, &hardlinkChange{
+					kind:      kind,
+					path:      path,
+					upperStat: upperStat,
+				})
+				hardlinkMu.Unlock()
+
+				return nil
+
+			default:
+				eg.Go(func() error {
+					appliedChange, err := local.WriteFile(egCtx, kind, path, upperStat, remote)
+					if err != nil {
+						return err
+					}
+					cachedResultsMu.Lock()
+					cachedResults = append(cachedResults, appliedChange)
+					cachedResultsMu.Unlock()
+					if cacheCtx != nil {
+						if err := cacheCtx.HandleChange(appliedChange.Result().kind, path, appliedChange.Result().stat, nil); err != nil {
+							return fmt.Errorf("failed to handle change in content hasher: %w", err)
+						}
+					}
+
+					return nil
+				})
+				return nil
+			}
+
+		case ChangeKindDelete:
+			appliedChange, err := local.RemoveAll(egCtx, path)
+			if err != nil {
+				return err
+			}
+			cachedResultsMu.Lock()
+			cachedResults = append(cachedResults, appliedChange)
+			cachedResultsMu.Unlock()
+			// no need to apply removals to the cacheCtx since it starts empty every Sync call.
+			return nil
+
+		case ChangeKindNone:
+			appliedChange, err := local.GetPreviousChange(egCtx, path, lowerStat)
+			if err != nil {
+				return err
+			}
+			cachedResultsMu.Lock()
+			cachedResults = append(cachedResults, appliedChange)
+			cachedResultsMu.Unlock()
+			if cacheCtx != nil {
+				if err := cacheCtx.HandleChange(appliedChange.Result().kind, path, appliedChange.Result().stat, nil); err != nil {
+					return fmt.Errorf("failed to handle change in content hasher: %w", err)
+				}
+			}
+
+			return nil
+
+		default:
+			return fmt.Errorf("unsupported change kind: %s", kind)
+		}
+	})
+
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+
+	for _, hardlink := range hardlinks {
+		appliedChange, err := local.Hardlink(ctx, hardlink.kind, hardlink.path, hardlink.upperStat)
+		if err != nil {
+			return nil, err
+		}
+		cachedResultsMu.Lock()
+		cachedResults = append(cachedResults, appliedChange)
+		cachedResultsMu.Unlock()
+		if cacheCtx != nil {
+			if err := cacheCtx.HandleChange(appliedChange.Result().kind, hardlink.path, appliedChange.Result().stat, nil); err != nil {
+				return nil, fmt.Errorf("failed to handle change in content hasher: %w", err)
+			}
+		}
+	}
+
+	if forParents {
+		// we created the parent dirs, nothing else to do now
+		return nil, nil
+	}
+
+	ctx, copySpan := newSpan(ctx, "copy")
+	defer copySpan.End()
```

Comment:
nit: would recommend using `telemetry.End(copySpan, func() error { return rerr })` pattern so the span fails if an error occurs from this point onward